lib/Ui/Tree.elm

```diff
@@ -0,0 +1,195 @@
+
+        viewNode n =
+            Styled.span [ css [ Css.display Css.inlineFlex, Css.flex Css.auto ] ] <|
+                case mRightAlignedText of
+                    Nothing ->
+                        [ Styled.text <| leftAlignedText n ]
+
+                    Just rightAlignedText ->
+                        [ Styled.span [ css [ Css.flexGrow (Css.num 1) ] ] [ Styled.text <| leftAlignedText n ]
+                        , Styled.text <| rightAlignedText n
+                        ]
+
+        viewTreeNode : Tree a -> Styled.Html msg
+        viewTreeNode t =
+            let
+                node =
+                    Tree.label t
+            in
+            Styled.li []
+                [ Styled.div
+                    [ labelStyle node
+                    , StyledEvents.onClick <| liftMsg (Select node)
+                    ]
+                  <|
+                    case Tree.children t of
+                        [] ->
+                            [ viewNode node ]
+
+                        _ ->
+                            [ Styled.span
+                                [ chevronStyle
+                                ]
+                                [ Icon.view Icon.chevronRight ]
+                            , viewNode node
+                            ]
+                ]
+
+        viewTree t =
+            case ( Zipper.parent t, Zipper.children t ) of
+                -- If we are on focused on a leaf we view the parent
+                ( Just parent, [] ) ->
+                    viewTree parent
+
+                -- If we are focused on a node with a parent and children we show a link to the parent and the children
+                ( Just _, children ) ->
+                    Styled.li
+                        [ labelStyle <| Zipper.label t
+                        , css <| TextStyle.toCssStyle TextStyle.label
+                        , StyledEvents.onClick <| liftMsg GoToParent
+                        ]
+                        [ Styled.span
+                            [ chevronStyle
+                            ]
+                            [ Icon.view Icon.chevronLeft ]
+                        , viewNode <| Zipper.label t
+                        ]
+                        :: List.map viewTreeNode children
+
+                -- If we are focused on the root node we only show the children
+                ( Nothing, children ) ->
+                    List.map viewTreeNode children
+
+        viewSearch s searchOn =
+            List.map viewTreeNode << searchTree (String.contains s << String.toLower << searchOn)
+    in
+    Styled.ul [ css <| Css.listStyleType Css.none :: TextStyle.toCssStyle TextStyle.body ] <|
+        case model.state of
+            Focus ->
+                viewTree model.tree
+
+            Search s searchOn ->
+                viewSearch s searchOn (Zipper.toTree model.tree)
+
+
+searchTree : (a -> Bool) -> Tree a -> List (Tree a)
+searchTree pred x =
```

Comment:
filter?

Reply:
Maybe. I have hard time thinking about what filtering a tree actually means. I think I would expect it to keep nodes given a predicate but still return a Tree a and not a List(Tree a) which this does.